cpp/src/arrow/array/builder_dict.h

```diff
@@ -282,6 +294,163 @@ class DictionaryBuilderBase : public ArrayBuilder {
     return indices_builder_.AppendEmptyValues(length);
   }
 
+  Status AppendScalar(const Scalar& scalar, int64_t n_repeats) override {
+    if (!scalar.type->Equals(type())) {
```

Comment:
Do we really want to do this check every append or should this be left to callers?